File: android/src/lib/android-connection-service-snapshot.ts
import type {
  AndroidConnectionServiceRoutePolicy,
  AndroidConnectionServiceTarget,
} from './android-connection-service-commands';

export type AndroidConnectionServiceState =
  | 'idle'
  | 'resolving-target'
  | 'connecting'
  | 'mux-ready'
  | 'channels-ready'
  | 'healthy'
  | 'backoff-reconnect'
  | 'authentication-error'
  | 'terminal-error';

export interface AndroidConnectionServiceChannelSnapshot {
  channelId: string;
  state: 'opening' | 'open' | 'closing' | 'closed';
}

export interface AndroidConnectionServiceError {
  code:
    | 'invalid-command'
    | 'transport'
    | 'heartbeat-timeout'
    | 'authentication'
    | 'terminal'
    | 'webrtc-not-supported';
  message: string;
}

export interface AndroidConnectionServiceSnapshot {
  state: AndroidConnectionServiceState;
  generation: string | null;
  target: AndroidConnectionServiceTarget | null;
  route: AndroidConnectionServiceRoutePolicy | null;
  channels: AndroidConnectionServiceChannelSnapshot[];
  lastHeartbeatAt: number | null;
  lastActivityAt: number | null;
  nextRetryAt: number | null;
  error: AndroidConnectionServiceError | null;
  muxReadyPayload: Record<string, unknown> | null;
}

export type AndroidConnectionServiceEvent =
  | { type: 'bind-target'; target: AndroidConnectionServiceTarget }
  | { type: 'set-route-policy'; policy: AndroidConnectionServiceRoutePolicy }
  | { type: 'release-target'; targetKey: string; reason: string }
  | { type: 'transport-opening'; generation: string }
  | { type: 'mux-ready'; generation: string; muxReadyPayload: Record<string, unknown> }
  | { type: 'channel-opened'; generation: string; channelId: string }
  | { type: 'channel-closed'; generation: string; channelId: string; reason: string }
  | { type: 'heartbeat-pong'; generation: string; at: number }
  | { type: 'server-activity'; generation: string; at: number }
  | { type: 'heartbeat-missed'; generation: string }
  | { type: 'transport-failure'; generation: string; message: string }
  | { type: 'authentication-failure'; generation: string; message: string }
  | { type: 'terminal-failure'; generation: string; message: string }
  | { type: 'webrtc-not-supported'; generation: string; message: string }
  | { type: 'reconnect-attempt'; generation: string; at: number };

export interface AndroidConnectionServiceStateMachineOptions {
  now?: () => number;
  heartbeatMissesBeforeReconnect?: number;
}

const EMPTY_SNAPSHOT: AndroidConnectionServiceSnapshot = {
  state: 'idle',
  generation: null,
  target: null,
  route: null,
  channels: [],
  lastHeartbeatAt: null,
  lastActivityAt: null,
  nextRetryAt: null,
  error: null,
  muxReadyPayload: null,
};

function readonlySnapshot(snapshot: AndroidConnectionServiceSnapshot): AndroidConnectionServiceSnapshot {
  return {
    ...snapshot,
    target: snapshot.target ? { ...snapshot.target } : null,
    route: snapshot.route ? { ...snapshot.route } : null,
    channels: snapshot.channels.map((channel) => ({ ...channel })),
    error: snapshot.error ? { ...snapshot.error } : null,
    muxReadyPayload: snapshot.muxReadyPayload ? { ...snapshot.muxReadyPayload } : null,
  };
}

function rejectGeneration(
  currentGeneration: string | null,
  event: Extract<AndroidConnectionServiceEvent, { generation: string }>,
) {
  return currentGeneration !== event.generation;
}

export function createAndroidConnectionServiceStateMachine(options: AndroidConnectionServiceStateMachineOptions = {}) {
  const now = options.now ?? Date.now;
  const heartbeatMissesBeforeReconnect = options.heartbeatMissesBeforeReconnect ?? 3;
  let snapshot: AndroidConnectionServiceSnapshot = { ...EMPTY_SNAPSHOT };
  let retiredGenerations = new Set<string>();
  let projectionAttached = true;
  let consecutiveHeartbeatMisses = 0;

  const dispatch = (event: AndroidConnectionServiceEvent): boolean => {
    if (event.type === 'bind-target') {
      const identicalTarget = snapshot.target?.targetKey === event.target.targetKey;
      if (identicalTarget && snapshot.target && snapshot.state !== 'idle') {
        return true;
      }
      snapshot = {
        ...snapshot,
        state: 'resolving-target',
        generation: null,
        target: { ...event.target },
        channels: [],
        lastHeartbeatAt: null,
        lastActivityAt: null,
        nextRetryAt: null,
        error: null,
      };
      retiredGenerations = new Set();
      consecutiveHeartbeatMisses = 0;
      return true;
    }
    if (event.type === 'set-route-policy') {
      const identicalPolicy = JSON.stringify(snapshot.route) === JSON.stringify(event.policy);
      if (identicalPolicy) {
        return true;
      }
      snapshot = {
        ...snapshot,
        route: { ...event.policy },
      };
      if (snapshot.generation && snapshot.state !== 'idle') {
        retiredGenerations.add(snapshot.generation);
        snapshot = {
          ...snapshot,
          state: 'resolving-target',
          generation: null,
          channels: [],
          lastHeartbeatAt: null,
          lastActivityAt: null,
          nextRetryAt: null,
          error: null,
        };
        consecutiveHeartbeatMisses = 0;
      }
      if (snapshot.state !== 'idle') {
        snapshot = { ...snapshot };
      }
      // Route policy is a control fact. The native owner will perform the
      // physical replacement; this projection only exposes the new desired
      // policy and its resulting resolving state.
      return true;
    }
    if (event.type === 'release-target') {
      snapshot = {
        ...EMPTY_SNAPSHOT,
        target: null,
        generation: null,
        route: snapshot.route ? { ...snapshot.route } : null,
      };
      retiredGenerations = new Set();
      consecutiveHeartbeatMisses = 0;
      return true;
    }
    if (event.type === 'transport-opening') {
      if (retiredGenerations.has(event.generation)) {
        return false;
      }
    } else if ('generation' in event) {
      if (rejectGeneration(snapshot.generation, event) || retiredGenerations.has(event.generation)) {
        return false;
      }
    }

    switch (event.type) {
      case 'transport-opening':
        if (snapshot.state !== 'resolving-target' && snapshot.state !== 'backoff-reconnect') {
          return false;
        }
        snapshot = {
          ...snapshot,
          state: 'connecting',
          generation: event.generation,
          muxReadyPayload: null,
          error: null,
          nextRetryAt: null,
        };
        consecutiveHeartbeatMisses = 0;
        return true;
      case 'mux-ready':
        if (snapshot.state !== 'connecting' && snapshot.state !== 'mux-ready') {
          return false;
        }
        snapshot = {
          ...snapshot,
          state: 'mux-ready',
          error: null,
          muxReadyPayload: { ...event.muxReadyPayload },
        };
        consecutiveHeartbeatMisses = 0;
        return true;
      case 'channel-opened':
        if (snapshot.state !== 'mux-ready' && snapshot.state !== 'channels-ready') {
          return false;
        }
        if (!snapshot.channels.some((channel) => channel.channelId === event.channelId)) {
          snapshot.channels.push({ channelId: event.channelId, state: 'open' });
        }
        snapshot = {
          ...snapshot,
          state: 'channels-ready',
          channels: [...snapshot.channels],
          lastActivityAt: now(),
          error: null,
        };
        return true;
      case 'channel-closed':
        snapshot = {
          ...snapshot,
          channels: snapshot.channels.map((channel) => (
            channel.channelId === event.channelId ? { ...channel, state: 'closed' } : channel
          )),
          lastActivityAt: now(),
        };
        return true;
      case 'heartbeat-pong':
        if (snapshot.state !== 'channels-ready' && snapshot.state !== 'mux-ready' && snapshot.state !== 'healthy') {
          return false;
        }
        snapshot = {
          ...snapshot,
          state: 'healthy',
          lastHeartbeatAt: event.at,
          lastActivityAt: event.at,
          error: null,
        };
        consecutiveHeartbeatMisses = 0;
        return true;
      case 'server-activity':
        snapshot = {
          ...snapshot,
          lastActivityAt: event.at,
          error: null,
        };
        consecutiveHeartbeatMisses = 0;
        return true;
      case 'heartbeat-missed':
        if (snapshot.state !== 'channels-ready' && snapshot.state !== 'mux-ready' && snapshot.state !== 'healthy') {
          return false;
        }
        consecutiveHeartbeatMisses += 1;
        if (consecutiveHeartbeatMisses < heartbeatMissesBeforeReconnect) {
          return true;
        }
        consecutiveHeartbeatMisses = 0;
        retiredGenerations.add(event.generation);
        snapshot = {
          ...EMPTY_SNAPSHOT,
          target: snapshot.target ? { ...snapshot.target } : null,
          route: snapshot.route ? { ...snapshot.route } : null,
          state: 'backoff-reconnect',
          generation: null,
          nextRetryAt: now() + 1000,
          lastActivityAt: now(),
          error: { code: 'heartbeat-timeout', message: 'mux heartbeat budget exhausted' },
        };
        return true;
      case 'transport-failure':
        retiredGenerations.add(event.generation);
        consecutiveHeartbeatMisses = 0;
        snapshot = {
          ...EMPTY_SNAPSHOT,
          target: snapshot.target ? { ...snapshot.target } : null,
          route: snapshot.route ? { ...snapshot.route } : null,
          state: 'backoff-reconnect',
          generation: null,
          nextRetryAt: now() + 1000,
          lastActivityAt: now(),
          error: { code: 'transport', message: event.message },
        };
        return true;
      case 'authentication-failure':
        retiredGenerations.add(event.generation);
        snapshot = {
          ...snapshot,
          state: 'authentication-error',
          generation: null,
          nextRetryAt: null,
          error: { code: 'authentication', message: event.message },
        };
        return true;
      case 'terminal-failure':
        retiredGenerations.add(event.generation);
        snapshot = {
          ...snapshot,
          state: 'terminal-error',
          generation: null,
          nextRetryAt: null,
          error: { code: 'terminal', message: event.message },
        };
        return true;
      case 'webrtc-not-supported':
        retiredGenerations.add(event.generation);
        snapshot = {
          ...snapshot,
          state: 'terminal-error',
          generation: null,
          nextRetryAt: null,
          error: { code: 'webrtc-not-supported', message: event.message },
        };
        return true;
      case 'reconnect-attempt':
        if (snapshot.state !== 'backoff-reconnect') {
          return false;
        }
        snapshot = {
          ...snapshot,
          state: 'resolving-target',
          nextRetryAt: null,
          error: null,
        };
        return true;
      default:
        return false;
    }
  };

  return {
    dispatch,
    readSnapshot: () => readonlySnapshot(snapshot),
    detachProjection: () => {
      projectionAttached = false;
    },
    attachProjection: () => {
      projectionAttached = true;
      return readonlySnapshot(snapshot);
    },
    isProjectionAttached: () => projectionAttached,
  };
}

File: android/native/android/app/src/main/java/com/zterm/android/AndroidConnectionServiceSnapshot.java
package com.zterm.android;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

/** Immutable UI projection of AndroidConnectionService state. */
public final class AndroidConnectionServiceSnapshot {
    public enum State {
        IDLE("idle"),
        RESOLVING_TARGET("resolving-target"),
        CONNECTING("connecting"),
        MUX_READY("mux-ready"),
        CHANNELS_READY("channels-ready"),
        HEALTHY("healthy"),
        BACKOFF_RECONNECT("backoff-reconnect"),
        AUTHENTICATION_ERROR("authentication-error"),
        TERMINAL_ERROR("terminal-error");

        private final String wireName;
        State(String wireName) { this.wireName = wireName; }
        public String wireName() { return wireName; }
    }

    public static final class Channel {
        public enum State {
            OPENING("opening"), OPEN("open"), CLOSING("closing"), CLOSED("closed");
            private final String wireName;
            State(String wireName) { this.wireName = wireName; }
            public String wireName() { return wireName; }
        }

        public final String channelId;
        public final State state;

        public Channel(String channelId, State state) {
            this.channelId = channelId;
            this.state = state;
        }

        public JSONObject toJson() throws JSONException {
            JSONObject json = new JSONObject();
            json.put("channelId", channelId);
            json.put("state", state.wireName());
            return json;
        }
    }

    public static final class ErrorValue {
        public final String code;
        public final String message;

        public ErrorValue(String code, String message) {
            this.code = code;
            this.message = message;
        }

        public JSONObject toJson() throws JSONException {
            JSONObject json = new JSONObject();
            json.put("code", code);
            json.put("message", message);
            return json;
        }
    }

    public final State state;
    public final String generation;
    public final AndroidConnectionServiceTarget target;
    public final AndroidConnectionServiceRoutePolicy route;
    public final List<Channel> channels;
    public final Long lastHeartbeatAt;
    public final Long lastActivityAt;
    public final Long nextRetryAt;
    public final ErrorValue error;
    public final String muxReadyPayloadJson;

    private AndroidConnectionServiceSnapshot(Builder b) {
        this.state = b.state;
        this.generation = b.generation;
        this.target = b.target;
        this.route = b.route;
        this.channels = Collections.unmodifiableList(new ArrayList<>(b.channels));
        this.lastHeartbeatAt = b.lastHeartbeatAt;
        this.lastActivityAt = b.lastActivityAt;
        this.nextRetryAt = b.nextRetryAt;
        this.error = b.error;
        this.muxReadyPayloadJson = b.muxReadyPayloadJson;
    }

    public static AndroidConnectionServiceSnapshot empty() {
        return new Builder(State.IDLE).build();
    }

    public static AndroidConnectionServiceSnapshot emptyForTarget(AndroidConnectionServiceTarget target) {
        return new Builder(State.IDLE).target(target).build();
    }

    public Builder toBuilder() {
        return new Builder(state)
            .generation(generation)
            .target(target)
            .route(route)
            .channels(channels)
            .lastHeartbeatAt(lastHeartbeatAt)
            .lastActivityAt(lastActivityAt)
            .nextRetryAt(nextRetryAt)
            .error(error)
            .muxReadyPayloadJson(muxReadyPayloadJson);
    }

    public JSONObject toJson() throws JSONException {
        JSONObject json = new JSONObject();
        json.put("state", state.wireName());
        json.put("generation", generation == null ? JSONObject.NULL : generation);
        json.put("target", target == null ? JSONObject.NULL : target.toJson());
        json.put("route", route == null ? JSONObject.NULL : route.toJson());
        JSONArray channelArray = new JSONArray();
        for (Channel channel : channels) channelArray.put(channel.toJson());
        json.put("channels", channelArray);
        json.put("lastHeartbeatAt", lastHeartbeatAt == null ? JSONObject.NULL : lastHeartbeatAt);
        json.put("lastActivityAt", lastActivityAt == null ? JSONObject.NULL : lastActivityAt);
        json.put("nextRetryAt", nextRetryAt == null ? JSONObject.NULL : nextRetryAt);
        json.put("error", error == null ? JSONObject.NULL : error.toJson());
        json.put("muxReadyPayload", muxReadyPayloadJson == null
            ? JSONObject.NULL : new JSONObject(muxReadyPayloadJson));
        return json;
    }

    @Override
    public boolean equals(Object other) {
        if (!(other instanceof AndroidConnectionServiceSnapshot)) return false;
        AndroidConnectionServiceSnapshot that = (AndroidConnectionServiceSnapshot) other;
        return state == that.state
            && Objects.equals(generation, that.generation)
            && Objects.equals(target, that.target)
            && Objects.equals(route == null ? null : route.mode, that.route == null ? null : that.route.mode)
            && Objects.equals(route == null ? null : route.path, that.route == null ? null : that.route.path)
            && channelsEqual(channels, that.channels)
            && Objects.equals(lastHeartbeatAt, that.lastHeartbeatAt)
            && Objects.equals(lastActivityAt, that.lastActivityAt)
            && Objects.equals(nextRetryAt, that.nextRetryAt)
            && Objects.equals(error == null ? null : error.code, that.error == null ? null : that.error.code)
            && Objects.equals(error == null ? null : error.message, that.error == null ? null : that.error.message)
            && Objects.equals(muxReadyPayloadJson, that.muxReadyPayloadJson);
    }

    private static boolean channelsEqual(List<Channel> a, List<Channel> b) {
        if (a.size() != b.size()) return false;
        for (int i = 0; i < a.size(); i++) {
            if (!Objects.equals(a.get(i).channelId, b.get(i).channelId)
                || a.get(i).state != b.get(i).state) return false;
        }
        return true;
    }

    @Override
    public int hashCode() {
        return Objects.hash(state, generation, target, route == null ? null : route.mode,
            route == null ? null : route.path, channels.size(), lastHeartbeatAt,
            lastActivityAt, nextRetryAt, error == null ? null : error.code,
            error == null ? null : error.message, muxReadyPayloadJson);
    }

    public static final class Builder {
        private State state;
        private String generation;
        private AndroidConnectionServiceTarget target;
        private AndroidConnectionServiceRoutePolicy route;
        private List<Channel> channels = new ArrayList<>();
        private Long lastHeartbeatAt;
        private Long lastActivityAt;
        private Long nextRetryAt;
        private ErrorValue error;
        private String muxReadyPayloadJson;

        public Builder(State state) { this.state = state; }
        public Builder generation(String v) { this.generation = v; return this; }
        public Builder target(AndroidConnectionServiceTarget v) { this.target = v; return this; }
        public Builder route(AndroidConnectionServiceRoutePolicy v) { this.route = v; return this; }
        public Builder channels(List<Channel> v) { this.channels = v == null ? new ArrayList<>() : new ArrayList<>(v); return this; }
        public Builder lastHeartbeatAt(Long v) { this.lastHeartbeatAt = v; return this; }
        public Builder lastActivityAt(Long v) { this.lastActivityAt = v; return this; }
        public Builder nextRetryAt(Long v) { this.nextRetryAt = v; return this; }
        public Builder error(ErrorValue v) { this.error = v; return this; }
        public Builder muxReadyPayloadJson(String v) { this.muxReadyPayloadJson = v; return this; }
        public AndroidConnectionServiceSnapshot build() { return new AndroidConnectionServiceSnapshot(this); }
    }
}
